### ISSUE TITLE
Routing Errors When to_param is Overwritten

### ISSUE THREAD

**Issue by stungeye** [activeadmin/activeadmin#279]
I overwrote `to_param` in a model to [implement friendly urls](http://www.miguelsanmiguel.com/2011/03/17/slug-that-slash).  Active Admin can no longer link to the admin urls that require ids without throwing a routing error.

party.rb:

```ruby
# Should give me urls like this: /parties/23/name-of-party
def to_param
  "#{id}/#{slug}"
end

def slug
  self.name.parameterize
end
```

routes.rb:
```ruby
resources :parties, :constraints => { :id => /[0-9]+\/.+/ }
```

I believe this a different problem than the other `to_param` issues listed below as I'm getting a routing error, not a active record loading error. Specifically, I get a routing error when any active admin page attempts to build a link that requires an id.

Is there any way to force active admin to ignore the overwritten to_param method?

Other issue that involve to_param:

#35: https://github.com/gregbell/active_admin/issues/35
#62: https://github.com/gregbell/active_admin/issues/62
#191: https://github.com/gregbell/active_admin/issues/191

**Comment by lukerollans** [comment]
This works perfectly for me :+1: 

Thanks @mdoyle13 

```
controller do
    defaults :finder => :find_by_slug
end
```

**Comment by timoschilling** [comment]
I'm closing this, while it's a inherited_resources bug and not a ActiveAdmin one.

This should be the correct way to handle it:
```ruby
controller do
  defaults :finder => :find_by_slug
end
```

**Comment by aprescott** [comment]
I think there's more than one problem being discussed on this issue, and `defaults finder: :find_by_slug` only addresses one of them.

As I see it:

1. Some people have a custom `to_param` which they need to tie back into their ActiveAdmin pages, so that things can be looked up properly.
2. Some people have a custom `to_param` but don't want it to have any effect on ActiveAdmin pages, so that everything is based on ID.

(2) causes a bunch of problems. It means that you have to specify `.id` on any `link_to`s in the view templates, but it also means that for stuff like `index` blocks, you have to reimplement the link to "View" a particular record.

```ruby
column "" do |foo|
  link_to "View", admin_foo_path(foo.id)
end
```

That's very hacky and isn't very sustainable in the face of future changes to how ActiveAdmin renders the index table, or handling permissions for which links are visible by a particular admin user. (It's also fairly cringeworthy that other solutions posted here include dynamically changing method definitions on the fly.)

So basically while (1) is solved by specifying a `:finder`, I don't think (2) has been addressed. Does it not belong in this issue?

**Comment by timoschilling** [comment]
@aprescott If you use a custom `to_param` you need to use it everywhere. To build a `config.use_to_param = false` would be a massiv code blow up, that we can't do. Thats way we use the Rails url helpers under the hood and they use the `to_param` method. You have two choices, customize `to_param` or not. Customize `to_param`, but not for ActiveAdmin is not and probably will not be supported way.

Here are some DIY ideas:
1. You can use Rubys Refinements to redefine `url_for` in ActiveAdmin to use `.id` instate of `.to_params`.
2. You can work with ViewModels to have a `User`, a `FrontendUser` and a `BackendUser` class. Then you can define `to_params` only on the `FrontendUser`. (depend on what you wan't to do, you don't need all three classes)
3. Maybe a Decorator can help you.

BTW: It's not only a ActiveAdmin problem, we use Formtastic and inherited_resources which both need to change too.

**Comment by manuca** [comment]
I see this is an old thread but I wanted to comment on a solution that worked for me while hitting this same problem. What I did was make an AdminJob model that inherits from Job and defines to_param the standard way (return instance id). That is:

```ruby
class Job < ActiveRecord::Base
end

class AdminJob < Job
  def to_param
    id.to_s
  end
end
class 
```

Then on Active Admin's side I did:

```ruby
ActiveAdmin.register AdminJob, as: "Job" do
end
```

I currently haven't find any problems with this, it seems ActiveRecord is quite smart with inheritance.

Cheers

**Comment by tomsword** [comment]
+1 for mdoyle13 (UPDATE 2). Works for me.

**Comment by AlexVPopov** [comment]
`defaults finder: :find_by_slug` works, however, if I have a child `belongs_to` resource within the parent resource, I again get an error:

```ruby
Couldn't find Foo with 'id'=bar
```

This happens in:
```ruby
def scoped_collection
  super.includes(:bazes)
end
```
This method is defined within the `controller` block of the child resource. So I guess, there must be something like:

```
defaults parent_finder: :find_by_slug
```

**Comment by AlexVPopov** [comment]
OK, so it seems that the resources configurations for a parent resources are set by the options of the `belongs_to`. So in order to be able to use a custom finder for the parent resource, you need to do this:

```ruby
belongs_to :foo, finder: :find_by_slug
```
